server/api_conversation.go

```diff
@@ -0,0 +1,158 @@
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get bot: %w", err))
+		return
+	}
+
+	var reqData ConversationRequest
+	if err := json.NewDecoder(c.Request.Body).Decode(&reqData); err != nil {
+		c.AbortWithError(http.StatusBadRequest, err)
+		return
+	}
+	defer c.Request.Body.Close()
+
+	// Validation
+	if reqData.BotName == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty bot"))
+		return
+	}
+
+	bot := p.GetBotByUsername(reqData.BotName)
+	if bot == nil {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid bot name"))
+		return
+	}
+
+	post := reqData.Request
+
+	if post == nil {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid request"))
+		return
+	}
+
+	if post.Message == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty message"))
+		return
+	}
+
+	if post.ChannelId == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty channel id"))
+		return
+	}
+
+	channel, err := p.pluginAPI.Channel.Get(post.ChannelId)
+	if errors.Is(err, pluginapi.ErrNotFound) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("channel not found"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get channel: %w", err))
+		return
+	}
+
+	if post.UserId == "" {
+		c.AbortWithError(http.StatusBadRequest, errors.New("invalid empty user id"))
+		return
+	}
+
+	postingUser, err := p.pluginAPI.User.Get(post.UserId)
+	if errors.Is(err, pluginapi.ErrNotFound) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("user not found"))
+		return
+	} else if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get posting user: %w", err))
+		return
+	}
+
+	// Don't respond to ourselves
+	if p.IsAnyBot(post.UserId) {
+		c.AbortWithError(http.StatusBadRequest, errors.New("not responding to ourselves"))
+		return
+	}
+
+	list := &model.PostList{
+		Order: make([]string, 0, len(reqData.Thread)+1),
+		Posts: make(map[string]*model.Post, len(reqData.Thread)+1),
+	}
+	list.Order = append(list.Order, post.Id)
+	list.Posts[post.Id] = post
+	for i, post := range reqData.Thread {
+		list.Order = append(list.Order, post.Id)
+		list.Posts[post.Id] = reqData.Thread[i]
+	}
+
+	threadData, err := p.getMetadataForPosts(list)
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to get thread data: %w", err))
+		return
+	}
+
+	prompt, err := p.prompts.ChatCompletion(ai.PromptDirectMessageQuestion, p.MakeConversationContext(bot, postingUser, channel, post))
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, fmt.Errorf("failed to generate prompt: %w", err))
+		return
+	}
+	prompt.AppendConversation(p.ThreadToBotConversation(bot, threadData.Posts))
+
+	// Overriding post role if requested.
+	if reqData.UseSystemRole {
+		for i := range prompt.Posts {
+			prompt.Posts[i].Role = ai.PostRoleSystem
+		}
+	}
```

Comment:
A bit brutal but it gets the job done. I can turn this into a method of `BotConversation` and refine further of course.